web/src/dictionary/markup/to_jsx_child.ts

```diff
@@ -0,0 +1,111 @@
+import react_htmr from "htmr";
+import { JSX, VNode } from "preact";
+import render_to_string from "preact-render-to-string";
+import type { JSX_Child, Replacer } from "./types";
+
+/* WORKAROUND */
+
+const htmr = (input: string) => {
+    // htmr deals in React.ReactNode
+    // Preact deals in VNode
+    // So we cast to VNode as that's what we mostly use.
+    return react_htmr(input) as VNode;
+}
+
+/* TYPES */
+
+
+type Replacement = {
+    begin_index: number,
+    end_index: number,
+    jsx: JSX.Element,
+};
+
+// string.matchAll() requires a global regex.
+type Global_Regex = RegExp;
+
+
+/* ENTRY */
+
+
+export default function markup_to_jsx_child(
+    input: JSX_Child,
+    regex_string: string,
+    replacer: Replacer,
+    keep_children_as_string: boolean = false,
+) {
+    const string_input = jsx_children_to_string(input);
+
+    // We use a separate regex to test whether we need to make_replacements().
+    // This is because using regex.test() on a global regex advances its lastIndex
+    // and that will produce inconsistent behaviour.
+    if (!new RegExp(regex_string).test(string_input)) {
+        return input;
+    }
+    const replacements = make_replacements(
+        string_input,
+        new RegExp(regex_string, "g"),
+        replacer,
+        keep_children_as_string
+    );
+    return replace_string_with_jsx(string_input, replacements);
+}
+
+function jsx_children_to_string(input: JSX_Child): string {
+    if (typeof input === "string") {
+            return input;
+        }
+    return render_to_string(input);
+}
+
+function make_replacements(
+    input: string,
+    regex: Global_Regex,
+    replacer: Replacer,
+    keep_children_as_string: boolean
+) {
+    if (input.length === 0) {
+        return [];
+    }
+    const replacements: Replacement[] = [];
+    const matched = input.matchAll(regex);
+    for (const match_array of matched) {
+        const whole_string = match_array[0];
+        // Not every capture group of the regex will match. We filter out failed

```

Comment:
That doesn't sound right. It means in the replacer you could have variables get unexpected matching groups.
I would recommend keeping the matching intact and let the replacer code handle optional parts. This is consistent with most other regex apis.